### PR TITLE
Core, Spark: Fix migrate table in case of partitioned table with partition containing a special character

### PR DESCRIPTION
**Pull request by wypoon**
This is https://github.com/apache/iceberg/issues/7612 reported by @gaborkaszab.

In this change, we fix the Spark migrate procedure to be able to handle a partition where a partition value contains '/'.
We do this by fixing `TableMigrationUtil.listPartition`. Instead of constructing a partition key (a String like `a=1/b=2`) from the `Map<String, String>` of partition column to partition value that is passed in, and then calling `DataFiles.fillFromPath` (via `DataFiles.Builder#withPartitionPath`) with the partition key String, where it is then split along '/' and the value recovered for each partition column (by further splitting along '='), we simply construct a List of the partition values and hand that off.

(We retain `DataFiles.fillFromPath` and `DataFiles.Builder#withPartitionPath` for other existing callers.)

We add a unit test that shows the problem; without the fix, the test fails with 
```
java.lang.IllegalArgumentException: Invalid partition data, too many fields (expecting 1): data=2023/05/30
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:326)
	at org.apache.iceberg.DataFiles.fillFromPath(DataFiles.java:67)
	at org.apache.iceberg.DataFiles$Builder.withPartitionPath(DataFiles.java:246)
	at org.apache.iceberg.data.TableMigrationUtil.buildDataFile(TableMigrationUtil.java:190)
	at org.apache.iceberg.data.TableMigrationUtil.lambda$listPartition$3(TableMigrationUtil.java:131)
	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:413)
	at org.apache.iceberg.util.Tasks$Builder.runSingleThreaded(Tasks.java:219)
	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:203)
	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:196)
	at org.apache.iceberg.data.TableMigrationUtil.listPartition(TableMigrationUtil.java:126)
	at org.apache.iceberg.data.TableMigrationUtil.listPartition(TableMigrationUtil.java:83)
	at org.apache.iceberg.spark.SparkTableUtil.listPartition(SparkTableUtil.java:278)
```

### CODE REVIEW

**Comment by gaborkaszab** [comment]
Hey @wypoon , I have also uploaded a PR to fix this issues. Forgot to put it into the github issue, though. https://github.com/apache/iceberg/pull/7738
Anyway I believe DataFiles.Builder already has an interface to avoid using the path for providing a partition.

**Comment by wypoon** [comment]
@gaborkaszab thanks for reviewing and pointing out your PR, https://github.com/apache/iceberg/pull/7738. I haven't had a chance to review it yet, but I will do so.
@szehon-ho @RussellSpitzer would you mind reviewing this please? If https://github.com/apache/iceberg/pull/7738 is a better solution, I am happy to contribute the tests here to it.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java
Comment (wypoon) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:194]: I think Spark uses a `Map<String, String>` to encode the partition values to pass them to `TableMigrationUtil` because it doesn't have better information than that. https://github.com/apache/iceberg/blob/8858f1cfd3d1b0bd9b2d4230d1975a48d8bea1a8/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java#L173-L192
Spark's SessionCatalog returns a `org.apache.spark.sql.catalyst.catalog.CatalogTablePartition` which contains a `Map[String, String]`. So this is what `SparkTableUtil` has to work with.

File: core/src/main/java/org/apache/iceberg/DataFiles.java
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/DataFiles.java:99]: Could just do one check here size != size, "Invalid partition data, found %d fields, expected %d"

Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:99]: Sure, I'll just check for `partitionValues.size() == spec.fields().size()`. The code here was copied from `fillFromPath`, where there are two checks.
Btw, I think we need to use "%s" in the string template, as the guava code only looks for "%s", to replace with the varargs.

File: core/src/main/java/org/apache/iceberg/DataFiles.java
Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:271]: I don't see why not. `DataFiles.Builder` is essentially a helper class. If there is use for another helper method, I don't see why we shouldn't add it.

**Comment by wypoon** [comment]
@RussellSpitzer can you please review this again? I have addressed the feedback.

**Comment by gaborkaszab** [comment]
I made another round of review and I think this PR is fine.

**Review by gaborkaszab — APPROVED** [review]

**Comment by gaborkaszab** [comment]
Also comparing to the other solution, indeed this in-place approach seems better because it won't require any changes on the client side. One last thing I'd like to try is to make a build from this, load it under my local Impala and run my tests just to be on the safe side.

**Comment by gaborkaszab** [comment]
I pulled this fix into my env and ran my Impala tests on it. Everything works as expected! I think this patch is good to go. @RussellSpitzer would you mind taking a look? Impala/Spark/Hive would need this fix for table migration.

**Comment by wypoon** [comment]
@gaborkaszab thank you for reviewing and testing the fix with Impala!

@szehon-ho @flyrain can one of you help review this? Perhaps @RussellSpitzer is on vacation?

**Review by JonasJ-ap — APPROVED** [review]
LGTM! @wypoon Thanks for your contribution. Also, thanks @gaborkaszab for reporting this.

I have locally implemented the changes proposed in this PR and conducted some preliminary testing on the Delta Lake Migration module. It appears to address the same issue experienced in the delta lake migration module. Once this PR has been merged, I am happy to open a separate PR to let Delta Lake Migration module use `withPartitionValues` to resolve the issue.

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (zinking) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:107]: partitions ? minor

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (RussellSpitzer) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:218]: why is this renamed?

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (RussellSpitzer) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:107]: I'm confused about what happened here, are we modifying the exiting public method or adding a new one? We should attempt not to break compatibility if we can for public methods in utility classes.

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (RussellSpitzer) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:87]: Could we improve this description? It's not quite clear what this method is actually doing

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (RussellSpitzer) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:103]: number of partitions to search concurrently?

**Review by RussellSpitzer — COMMENTED** [review]
This PR is a bit challenging to review because it combines several cosmetic refactors with the actual code making the change. It would probably be much easier to read if we could separate out all the refactors that do not contribute to the core task of "handling partitions with a special character." Further changes to improve readability and so forth should go in a separate PR.

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (wypoon) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:107]: We are modifying an existing public method. There is no change to the signature of the method, although the names of a few parameters were changed to make them consistent with the other existing `listPartitions` method (which was actually the first `listPartitions` method). The difference between the first `listPartitions` and this second one is that this has an additional `parallelism` parameter. The javadoc for the first method is copied over to this second method with relevant addition (another reason to make the parameter names consistent).

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (wypoon) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:218]: Because the name is misleading. There are no deletes. This was obviously copied from somewhere else.
I shall remove this renaming.

**Comment by wypoon** [comment]
> This PR is a bit challenging to review because it combines several cosmetic refactors with the actual code making the change. It would probably be much easier to read if we could separate out all the refactors that do not contribute to the core task of "handling partitions with a special character." Further changes to improve readability and so forth should go in a separate PR.

I hear you. I shall remove the cosmetic refactors (except for renaming variables in `listPartitions`).

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (wypoon) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java]: You're right.

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java
Comment (wypoon) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java]: Actually no. I have improved the javadoc in the latest revision to explain the parallelism.
There is only one partition `listPartition` works on (as the name implies). The parallelism is in the reading of the data files (after the list of files is obtained) to get the metrics. The listing itself is done single-threaded.

**Comment by wypoon** [comment]
@RussellSpitzer thanks for your second review. I have improved the javadoc and removed unrelated refactors in `TableMigrationUtil`.

File: core/src/main/java/org/apache/iceberg/DataFiles.java
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/DataFiles.java:275]: This branch is never false

**Review by RussellSpitzer — APPROVED** [review]
One minor nit on the un-needed branch, I think this is good to go otherwise.

File: core/src/main/java/org/apache/iceberg/DataFiles.java
Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: Do you mean `partitionValues.isEmpty()` can never be true logically? Or it is never true because `withPartitionValues` is never called with an empty `partitionValues`?
If the latter, that is because I'm the only one who has written code calling `withPartitionValues` but there is nothing to stop future callers from calling it with empty `partitionValues`.
I copied the code from the existing `withPartitionPath`. Perhaps the `if` in both methods is overkill and protects against calls that don't exist (but in principle could occur).

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: This is actually a little more confusing still, "fillFromValues" throws an error if you have the wrong number of values so currently the logic looks like

1. isPartitioned + partitionValues.isEmpty - doesNothing
2. isPartitioned + parititonValues.size != partitonSpec.size && isNotEmpty - Error
3. isPartitioned + partitionValues.size == partitionSpec.size - does something
4. not partitioned + partitionvalues.isEmpty - doesNothing
5. not partitioned + parittionValues.nonEmpty - Error

In my opinion here everything but 3 should be illegal.

Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: I can change the `Preconditions.checkArgument` in `withPartitionValues` to
```
      Preconditions.checkArgument(
          isPartitioned && !partitionValues.isEmpty(),
          "Table must be partitioned and partition values must not be empty");
```
Then 1 and 4 will throw an error.
Only 3 will be allowed.

Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: Actually, I see now that we should leave things as they are.
`SparkTableUtil.importUnpartitionedSparkTable` actually does call `TableMigrationUtil.listPartition`, with an empty `Map` for `partition`, and so that will call `DataFiles.Builder#withPartitionValues` with an empty `List` for `partitionValues`. That should be allowed.

(I tried the change I just proposed above, and some tests in `TestMigrateTableProcedure` failed.)

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: So if it's unpartitioned then it must be empty. So either it is partitioned and values is as long as the spec. Or it is unpartitioned and empty?

Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: Yes, that is correct.
Right now we allow 1 and 4. We should allow 4, but to be strict, we should not allow 1. So, if we want to be strict, we can use
```
      Preconditions.checkArgument(
          (isPartitioned && !partitionValues.isEmpty())
              || (!isPartitioned && partitionValues.isEmpty()),
          "Table must be partitioned or partition values must be empty");
```
We still need the `if (!partitionValues.isEmpty())` though.

Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/DataFiles.java:275]: Or more succinctly,
```
      Preconditions.checkArgument(
          isPartitioned ^ partitionValues.isEmpty(),
          "Table must be partitioned or partition values must be empty");
```
I didn't realize that `^` could be used as a logical xor as well as bitwise xor.

**Comment by wypoon** [comment]
> One minor nit on the un-needed branch, I think this is good to go otherwise.

@RussellSpitzer I have addressed the minor nit as well.

**Comment by wypoon** [comment]
Can this be merged now? It's been a week.

**Comment by RussellSpitzer** [comment]
Merged, Thanks for your patience @wypoon I've been traveling this week so it's been very busy . Thanks also to @gaborkaszab and @JonasJ-ap for your review feedback.

**Pull request merged**
